src/tcp/TcpDuplexConnection.cpp

```diff
@@ -122,36 +116,113 @@ class TcpReaderWriter : public ::folly::AsyncTransportWrapper::WriteCallback,
     inputSubscriber_->onNext(std::move(readBuf));
   }
 
+  bool isClosed() const {
+    return !socket_;
+  }
+
+  void close() {
+    if (auto socket = std::move(socket_)) {
+      socket->close();
+    }
+    if (auto outputSubscription = std::move(outputSubscription_)) {
+      outputSubscription->cancel();
+    }
+  }
+
   folly::IOBufQueue readBuffer_{folly::IOBufQueue::cacheChainLength()};
   folly::AsyncSocket::UniquePtr socket_;
+  const std::shared_ptr<Stats> stats_;
 
   std::shared_ptr<reactivesocket::Subscriber<std::unique_ptr<folly::IOBuf>>>
       inputSubscriber_;
+  std::shared_ptr<reactivesocket::Subscription> outputSubscription_;
+};
+
+class TcpOutputSubscriber
+    : public SubscriberBaseT<std::unique_ptr<folly::IOBuf>> {
+ public:
+  TcpOutputSubscriber(
+      std::shared_ptr<TcpReaderWriter> tcpReaderWriter,
+      folly::Executor& executor)
+      : ExecutorBase(executor), tcpReaderWriter_(std::move(tcpReaderWriter)) {}
+
+  void onSubscribeImpl(
+      std::shared_ptr<Subscription> subscription) noexcept override {
+    if (tcpReaderWriter_) {
+      // no flow control at tcp level, since we can't know the size of messages
+      subscription->request(std::numeric_limits<size_t>::max());
```

Comment:
Is this necessary?  It seems like in the next line, `TcpReaderWriter::setOutputSubscription()` will also do `request(SIZE_T_MAX)`.